tripy/docs/pre0_user_guides/02-compiler.md

```diff
@@ -0,0 +1,170 @@
+```
+
+To naively run `GEGLU` in eager mode:
+
+```py
+# doc: no-eval
+layer = GEGLU(2, 8)
+inp = tp.ones((1, 2))
+out = layer(inp)
+print(out)
+```
+
+Now, let's try to optimize this model for inference using Tripy's Compiler.
+
+First, let's initialize the compiler with `layer` which lets the `Compiler` know the properties of the function we are compiling (e.g. function signature).
+
+```py
+# doc: no-eval
+compiler = tp.Compiler(layer)
+```
+Next, we need to provide information about each input using `tp.InputInfo`. The first argument for `InputInfo` is `shape`, where we specify either the static or dynamic shape information for each dimension. In the example below, we assume the shape of `inp` is static (`(1, 2)`). The second argument specifies the `dtype` for the input:
+
+```py
+# doc: no-eval
+inp_info = tp.InputInfo(shape=(1, 2), dtype=tp.float32)
+```
+Now, we can call the `compile` function to obtain a compiled function and use it for inference:
+
+```py
+# doc: no-eval
+fast_geglu = compiler.compile(inp_info)
+print(fast_geglu(inp))
+```
+
+### Optimization profiles
+
+In the example above, we assumed `inp` has a static shape of `(1, 2)`. Now, let’s assume that the shape of `inp` can vary from `(1, 2)` to `(16, 2)`. To express this constraint to the compiler, we can provide the range of shapes to `InputInfo` using `shape=((1, 8, 16), 2)`. This indicates to the compiler that the first dimension can vary from 1 to 16, and it should optimize for performance based on 8.
+
+```py
+# doc: no-eval
+inp_info = tp.InputInfo(shape=((1, 8, 16), 2), dtype=tp.float32)
+fast_geglu = compiler.compile(inp_info)
+print(fast_geglu(inp))
+
+# Let's change the shape of input to (2, 2)
+inp = tp.Tensor([[1., 2.], [2., 3.]], dtype=tp.float32)
+print(fast_geglu(inp))
+```
+
+### Errors for out of bounds inference
+If the user provides input that does not comply with the dynamic shape constraint given to the compiler, `Tripy` will produce an error with relevant information:
+
+```py
+# doc: no-eval
+from tests import helper
+
+inp = tp.ones((32, 2), dtype=tp.float32)
+def test_oob(self):
+        with helper.raises(tp.TripyException, match="Unexpected tensor shape."):
+            fast_geglu(inp).eval()
+
+# E       Unexpected tensor shape.
+# E           For parameter `x`, expected the tensor shape `shape(32, 2)` to be within bounds for all dimensions. However, dimension 0 has a shape of 32, which is not within the expected bounds of (1, 16). Note: The provided argument was: 
+# E       
+# E           --> /tripy/tripy/frontend/trace/ops/fill.py:121 in full()
+# E                 |
+# E             121 |     return full_impl(shape, value, dtype, output_rank)
+# E                 | 
+# E       
+# E           --> /tripy/tripy/frontend/ops/tensor_initializers.py:57 in ones()
+# E                 |
+# E              57 |     return full(shape, 1, dtype)
+# E                 |            ^^^^^^^^^^^^^^^^^^^^^ --- required from here
+# E       
+```
```

Comment:
Is is possible to render the error message in a dynamic way?

Reply:
Done in the latest iteration.